### PR TITLE
smoothie board repro

### PR DESCRIPTION
**Pull request by Abse2001**
Now instead of throwing errors when a `processPin` and `processVis`  fail to parse a node correctly, we are throwing a warning in the console and the note that did not parse correctly.

### CODE REVIEW

**Comment by Abse2001** [comment]
@seveibar is this how we handle these errors?

File: lib/dsn-pcb/dsn-json-to-circuit-json/parse-dsn-to-dsn-json.ts [hunk 1]
```diff
@@ -566,25 +566,36 @@ function processOutline(nodes: ASTNode[]): Outline {
...
 function processPin(nodes: ASTNode[]): Pin {
   const pin: Partial<Pin> = {}
+  // default pin
+  pin.padstack_name = "default"
+  pin.pin_number = 1
+  pin.x = 0
+  pin.y = 0
 
   try {
+    // Get padstack name
```
Comment (seveibar) [inline lib/dsn-pcb/dsn-json-to-circuit-json/parse-dsn-to-dsn-json.ts:576]:
```suggestion
```

**Comment by Abse2001** [comment]
@seveibar, how can I make the processPIn and processVia return a null without breaking any fn that uses it? Should I just change the types of any fn that needs to handle null, or is there a better way?
image

**Comment by Abse2001** [comment]
@seveibar I don't know why the Bun Test is failing even before my PR what do you think of the code

File: lib/dsn-pcb/dsn-json-to-circuit-json/parse-dsn-to-dsn-json.ts [hunk 2]
```diff
@@ -564,27 +565,34 @@ function processOutline(nodes: ASTNode[]): Outline {
...
+    }
+
+    // Helper function to parse pin number
+    const parsePinNumber = (value: any): number | string => {
+      if (typeof value === "number") {
+        return value
+      }
+      const parsed = parseInt(String(value), 10)
+      return Number.isNaN(parsed) ? String(value) : parsed
     }
```
Comment (seveibar) [inline lib/dsn-pcb/dsn-json-to-circuit-json/parse-dsn-to-dsn-json.ts:588]: this function has no deps, so it should not be contained inside another function. I would rename it and move it to a util directory.

It is also very strange that this function returns either a string or a number.

Create a util file and add function:

```ts
export const parseNumberOrReturnOriginalValue = (value: string | number) => {
   // ...
}
```

also remove the ternary and replace with the "early return" pattern. Ternaries are often confusing- if you're in a function already always prefer the early return. This function was hard for me to figure out, e.g. I don't know why you cast the value to a string so much (i still don't really understand why- but it was hard to read to make sure it was relatively harmless (?))

Comment (seveibar) [reply lib/dsn-pcb/dsn-json-to-circuit-json/parse-dsn-to-dsn-json.ts:588]: actually use `parseNumberOrNull(value) ?? value`, that's much more readable

you can define that function if it doesn't exist

File: lib/dsn-pcb/dsn-json-to-circuit-json/parse-dsn-to-dsn-json.ts [hunk 3]
```diff
@@ -818,14 +826,15 @@ export function processNetwork(nodes: ASTNode[]): Network {
...
 }
 
 function processNet(nodes: ASTNode[]): Net {
   const net: Partial<Net> = {}
-  if (nodes[1].type === "Atom" && typeof nodes[1].value === "string") {
-    net.name = nodes[1].value
+  // in the smoothie board some of the nets names are numbers like 3.3 so I need to convert them to strings
+  if (nodes[1].type === "Atom") {
+    net.name = nodes[1].value?.toString()
+    debug("net name was not string before parsing", net.name)
```
Comment (seveibar) [inline lib/dsn-pcb/dsn-json-to-circuit-json/parse-dsn-to-dsn-json.ts:837]: i don't understand this message- you're always executing it so i think it's misleading

Comment (seveibar) [reply lib/dsn-pcb/dsn-json-to-circuit-json/parse-dsn-to-dsn-json.ts:837]: you should be checking if it's not a string then emitting the `debug("warning: net name was not a string", net.name)`

File: lib/dsn-pcb/dsn-json-to-circuit-json/parse-dsn-to-dsn-json.ts [hunk 4]
```diff
@@ -950,32 +959,21 @@ function processVia(nodes: ASTNode[]): Wire {
...
-
-      return wire as Wire
+      wire.path!.coordinates = coords.map((node) => node.value as number)
+    } else {
+      debug("Warning: Missing or incomplete coordinates for via", nodes)
+      return null
     }
+  } else {
+    debug("Warning: Missing path node for via", nodes)
+    return null
```
Comment (seveibar) [inline lib/dsn-pcb/dsn-json-to-circuit-json/parse-dsn-to-dsn-json.ts:972]: switch this to an early return pattern, avoid long "if-else" statements, the early return pattern avoids them

File: lib/dsn-pcb/dsn-json-to-circuit-json/parse-dsn-to-dsn-json.ts [hunk 5]
```diff
@@ -950,32 +959,21 @@ function processVia(nodes: ASTNode[]): Wire {
...
-          node.children[0].value === "net",
-      )
-      if (netNode?.children?.[1]?.type === "Atom") {
-        wire.net = String(netNode.children[1].value)
-      }
-
-      return wire as Wire
+      wire.path!.coordinates = coords.map((node) => node.value as number)
+    } else {
+      debug("Warning: Missing or incomplete coordinates for via", nodes)
```
Comment (seveibar) [inline lib/dsn-pcb/dsn-json-to-circuit-json/parse-dsn-to-dsn-json.ts:967]: switch this to an early return pattern, avoid long "if-else" statements, the early return pattern avoids them

**Review by seveibar — CHANGES_REQUESTED** [review]
make sure to fix the type issues before requesting

File: lib/dsn-pcb/dsn-json-to-circuit-json/parse-dsn-to-dsn-json.ts (hunk 2 above)
Comment (Abse2001) [reply lib/dsn-pcb/dsn-json-to-circuit-json/parse-dsn-to-dsn-json.ts:588]: > this function has no deps, so it should not be contained inside another function. I would rename it and move it to a util directory.
> 
> It is also very strange that this function returns either a string or a number.
> 
> Create a util file and add function:
> 
> ```ts
> export const parseNumberOrReturnOriginalValue = (value: string | number) => {
>    // ...
> }
> ```
> 
> also remove the ternary and replace with the "early return" pattern. Ternaries are often confusing- if you're in a function already always prefer the early return. This function was hard for me to figure out, e.g. I don't know why you cast the value to a string so much (i still don't really understand why- but it was hard to read to make sure it was relatively harmless (?))

for this one, I think I forgot to remove it.

**Review by seveibar — APPROVED** [review]
NICE /tip $5

**Comment by seveibar** [comment]
@Abse2001 so now whenever we build major enhancements we should be able to see the result on the smoothie board, pretty cool! It basically represents an "End to End Test" that tests lots of stuff

**Pull request merged**